File: lab-02/src/modules/task_solutions.cpp
#include "task_solutions.h"

// Проверка сбалансированности скобок с помощью стека
bool check_brackets(const std::string& s) {
    std::vector<char> stack;
    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') stack.push_back(c);
        else if (c == ')' || c == ']' || c == '}') {
            if (stack.empty()) return false;
            char top = stack.back();
            stack.pop_back();
            if ((c == ')' && top != '(') ||
                (c == ']' && top != '[') ||
                (c == '}' && top != '{')) return false;
        }
    }
    return stack.empty();
}

// Симуляция очереди печати
void print_queue_simulation() {
    std::deque<std::string> queue;
    queue.push_back("Документ1");
    queue.push_back("Документ2");
    queue.push_back("Документ3");

    std::cout << "=== Очередь печати ===\n";
    while (!queue.empty()) {
        std::cout << "Печатается: " << queue.front() << std::endl;
        queue.pop_front();
    }
}

// Проверка палиндрома
bool is_palindrome(const std::string& s) {
    std::deque<char> dq;
    for (char c : s)
        if (std::isalnum(c))
            dq.push_back(std::tolower(c));

    while (dq.size() > 1) {
        if (dq.front() != dq.back()) return false;
        dq.pop_front();
        dq.pop_back();
    }
    return true;
}

File: lab-02/src/modules/task_solutions.py
from collections import deque


def check_brackets(s):
    """
    Проверка сбалансированности скобок.
    Используется стек на списке — операции push/pop O(1).
    Общая сложность: O(n).
    """
    stack = []  # O(1)

    for c in s:  # O(n)
        if c in "([{":           # O(1)
            stack.append(c)      # O(1)
        elif c in ")]}":         # O(1)
            if not stack:        # O(1)
                # Итоговая сложность функции check_brackets: O(n)
                return False     # O(1)

            top = stack.pop()    # O(1)

            if (c == ")" and top != "(") or \
               (c == "]" and top != "[") or \
               (c == "}" and top != "{"):   # O(1)
                # Итоговая сложность функции check_brackets: O(n)
                return False                # O(1)

    # Итоговая сложность функции check_brackets: O(n)
    return len(stack) == 0  # O(1)


def print_queue_simulation():
    """
    Симуляция работы очереди печати.
    deque обеспечивает операции push/pop O(1).
    """
    queue = deque()                  # O(1)
    queue.append("Документ1")        # O(1)
    queue.append("Документ2")        # O(1)
    queue.append("Документ3")        # O(1)

    print("=== Очередь печати ===")
    while queue:                     # O(n)
        print(f"Печатается: {queue.popleft()}")  # popleft — O(1)
    # Итоговая сложность функции print_queue_simulation: O(n)


def is_palindrome(s):
    """
    Проверка, является ли строка палиндромом.
    Используется deque, т.к. pop с обоих концов — O(1).
    Общая сложность: O(n).
    """
    dq = deque()  # O(1)

    for c in s:   # O(n)
        if c.isalnum():                    # O(1)
            dq.append(c.lower())           # O(1)

    while len(dq) > 1:                     # O(n)
        if dq.popleft() != dq.pop():       # каждая операция O(1)
            # Итоговая сложность функции is_palindrome: O(n)
            return False                   # O(1)

    # Итоговая сложность функции is_palindrome: O(n)
    return True  # O(1)
